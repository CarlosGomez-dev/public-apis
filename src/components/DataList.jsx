import './DataList.scss';

const headers = ['Name', 'Description', 'Auth', 'HTTPS', 'Category'];

export const DataList = ({ apiData, handleSort }) => (
  <table className='api-table'>
    <thead>
      <tr>
        {headers.map(header => (
          <th key={header}>
            <button
              className='api-table-header'
              onClick={() => handleSort(header === 'Name' ? 'API' : header)}
            >
              {header}
            </button>
          </th>
        ))}
      </tr>
    </thead>
    <tbody>
      {apiData.map(api => (
        <tr key={`${api.API}${api.Link}`}>
          <td className='api-name'>
            <a href={api.Link} target='_blank' rel='noopener noreferrer'>
              <span>{api.API}</span>
            </a>
          </td>
          <td className='api-description'>
            <span>{api.Description}</span>
          </td>
          <td>
            <span className={'badge ' + (api.Auth ? 'auth' : 'no-auth')}>
              {api.Auth || 'No Auth'}
            </span>
          </td>
          <td>
            <span className={'badge ' + (api.HTTPS ? 'https' : 'no-https')}>
              HTTP{api.HTTPS ? 'S' : ''}
            </span>
          </td>
          <td>
            <span className='badge category'>{api.Category}</span>
          </td>
        </tr>
      ))}
    </tbody>
  </table>
);
