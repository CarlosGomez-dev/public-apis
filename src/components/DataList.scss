@import '../colors';

.api-table {
  width: 100%;
  margin: 0 auto;
  border-collapse: collapse;
  thead tr {
    display: none;
    user-select: none;
  }
  tbody tr {
    display: flex;
    flex-wrap: wrap;
    padding: 0.5em;
    margin-bottom: 0.5em;
  }
  td,
  th {
    padding: 0.3rem;
    span {
      margin: 0.5rem 0;
    }
  }
}

.api-table-header {
  cursor: pointer;
  background-color: $background-color-table;
  color: $font-color;
  border: none;
  font-size: 1rem;
  font-weight: 600;
  padding: 0.5rem 0;
  display: flex;
  &::after {
    content: url('../assets/arrows-sort.svg');
    display: inline-block;
    width: 16px;
    margin: 0 3px;
  }
}

.api-name {
  width: 100%;
  font-size: 1.5rem;
  font-weight: 600;
  a {
    color: $font-color;
    &::after {
      content: url('../assets/external-link.svg');
      display: inline-block;
      width: 20px;
      margin: 0 1px;
    }
  }
}

.api-description {
  margin-top: 0.5rem;
  font-size: 1.3rem;
  order: 1;
  width: 100%;
  &::before {
    content: open-quote;
  }
  &::after {
    content: close-quote;
  }
}

.badge {
  border-radius: 6px;
  font-size: 0.7rem;
  padding: 0.3rem 0.4rem;
  color: $font-color-inverse;
  white-space: nowrap;
}
.category {
  background-color: $badge-color-neutral;
}
.no-auth,
.https {
  background-color: $badge-color-blue;
}
.auth,
.no-https {
  background-color: $badge-color-red;
}

@media (min-width: 768px) {
  .api-table {
    thead tr,
    tbody tr {
      display: table-row;
      background-color: $background-color-table;
      border: 1px solid $border-color;
      text-align: left;
    }
  }

  .api-name,
  .api-description {
    width: auto;
    font-size: clamp(0.8rem, 2vw, 1.1rem);
  }

  .api-name {
    a {
      text-decoration-color: transparent;
      transition: text-decoration-color 250ms;
      &:hover {
        text-decoration-color: $font-color;
      }
      &::after {
        content: url('../assets/external-link.svg');
        display: inline-block;
        width: 16px;
        margin: 0;
      }
    }
  }

  .auth {
    display: block;
    width: fit-content;
  }
  .category {
    display: block;
    text-align: center;
    white-space: unset;
  }
}
