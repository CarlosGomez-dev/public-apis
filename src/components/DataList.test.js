import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { DataList } from './DataList';

const apiData = [
  {
    API: 'Cat Facts',
    Description: 'Daily cat facts',
    Auth: '',
    HTTPS: true,
    Cors: 'no',
    Link: 'https://alexwohlbruck.github.io/cat-facts/',
    Category: 'Animals',
  },
  {
    API: 'Cataas',
    Description: 'Cat as a service (cats pictures and gifs)',
    Auth: '',
    HTTPS: false,
    Cors: 'unknown',
    Link: 'https://cataas.com/',
    Category: 'Animals',
  },
  {
    API: 'catAPI',
    Description: 'Random pictures of cats',
    Auth: 'apiKey',
    HTTPS: true,
    Cors: 'yes',
    Link: 'https://github.com/ThatCopy/catAPI/wiki/Usage',
    Category: 'Animals',
  },
];

const handleSort = jest.fn();

describe('Data List', () => {
  it('matches snapshot', () => {
    const { container } = render(<DataList apiData={apiData} handleSort={handleSort} />);
    expect(container).toMatchSnapshot();
  });
  it.each(apiData)('contains information for API %#', api => {
    render(<DataList apiData={apiData} handleSort={handleSort} />);
    const apiName = screen.getByText(api.API);
    const parentLink = apiName.parentElement;
    expect(screen.getByText(api.Description)).toBeInTheDocument();
    expect(parentLink.getAttribute('href')).toBe(api.Link);
  });
  it('contains Auth and HTTP badges', () => {
    render(<DataList apiData={apiData} handleSort={handleSort} />);
    const noAuthCount = apiData.reduce((acc, api) => (api.Auth === '' ? acc + 1 : acc), 0);
    const httpsCount = apiData.reduce((acc, api) => (api.HTTPS ? acc + 1 : acc), 1);
    expect(screen.getAllByText('No Auth').length).toBe(noAuthCount);
    expect(screen.getAllByText('HTTPS').length).toBe(httpsCount);
  });
  it('contains clickable headings', () => {
    render(<DataList apiData={apiData} handleSort={handleSort} />);
    const headers = screen.getAllByRole('columnheader');
    userEvent.click(headers[0].firstElementChild);
    expect(handleSort).toBeCalledTimes(1);
  });
});
