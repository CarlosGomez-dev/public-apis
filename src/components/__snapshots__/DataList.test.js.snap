// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data List matches snapshot 1`] = `
<div>
  <main
    class="card-list"
  >
    <article
      class="card"
    >
      <h2>
        <span>
          Cat Facts
        </span>
      </h2>
      <div
        class="card-badges"
      >
        <p
          class="badge no-auth"
        >
          <span>
            No Auth
          </span>
        </p>
        <p
          class="badge https"
        >
          <span>
            HTTP
            S
          </span>
        </p>
        <p
          class="badge category"
          style="background-color: rgb(42, 248, 92); color: rgb(0, 0, 0);"
        >
          Animals
        </p>
      </div>
      <p
        class="card-description"
      >
        Daily cat facts
      </p>
      <a
        class="card-link"
        href="https://alexwohlbruck.github.io/cat-facts/"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://alexwohlbruck.github.io/cat-facts/
      </a>
    </article>
    <article
      class="card"
    >
      <h2>
        <span>
          Cataas
        </span>
      </h2>
      <div
        class="card-badges"
      >
        <p
          class="badge no-auth"
        >
          <span>
            No Auth
          </span>
        </p>
        <p
          class="badge no-https"
        >
          <span>
            HTTP
            
          </span>
        </p>
        <p
          class="badge category"
          style="background-color: rgb(42, 248, 92); color: rgb(0, 0, 0);"
        >
          Animals
        </p>
      </div>
      <p
        class="card-description"
      >
        Cat as a service (cats pictures and gifs)
      </p>
      <a
        class="card-link"
        href="https://cataas.com/"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://cataas.com/
      </a>
    </article>
    <article
      class="card"
    >
      <h2>
        <span>
          catAPI
        </span>
      </h2>
      <div
        class="card-badges"
      >
        <p
          class="badge auth"
        >
          <span>
            apiKey
          </span>
        </p>
        <p
          class="badge https"
        >
          <span>
            HTTP
            S
          </span>
        </p>
        <p
          class="badge category"
          style="background-color: rgb(42, 248, 92); color: rgb(0, 0, 0);"
        >
          Animals
        </p>
      </div>
      <p
        class="card-description"
      >
        Random pictures of cats
      </p>
      <a
        class="card-link"
        href="https://github.com/ThatCopy/catAPI/wiki/Usage"
        rel="noopener noreferrer"
        target="_blank"
      >
        https://github.com/ThatCopy/catAPI/wiki/Usage
      </a>
    </article>
  </main>
</div>
`;
